File: src/main/java/org/polyfrost/hytils/handlers/chat/modules/blockers/ShoutBlocker.java
/*
 * Hytils Reborn - Hypixel focused Quality of Life mod.
 * Copyright (C) 2020, 2021, 2022, 2023  Polyfrost, Sk1er LLC and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.polyfrost.hytils.handlers.chat.modules.blockers;

import dev.deftu.textile.minecraft.MCTextHolder;
import net.hypixel.data.type.GameType;
import net.minecraft.util.IChatComponent;
import org.polyfrost.hytils.config.HytilsConfig;
import org.polyfrost.hytils.handlers.chat.ChatReceiveModule;
import org.polyfrost.hytils.handlers.chat.ChatSendModule;
import net.minecraft.client.Minecraft;
import org.polyfrost.oneconfig.api.event.v1.events.ChatEvent;
import org.jetbrains.annotations.NotNull;
import org.polyfrost.oneconfig.api.hypixel.v1.HypixelUtils;

import java.text.DecimalFormat;

/**
 * Blocks using /shout in Bedwars and Skywars modes, where there are multiple people in 1 team and also in Capture The Wool.
 * If there are more modes supporting /shout, feel free to add support for them.
 */
public class ShoutBlocker implements ChatSendModule, ChatReceiveModule {
    @Override
    public int getPriority() {
        return -3;
    }

    private long shoutCooldown = 0L;
    private final DecimalFormat decimalFormat = new DecimalFormat("#.#"); // only 1 decimal

    @Override
    public String onMessageSend(@NotNull String message) {
        if (message.startsWith("/shout ")) {
            if (shoutCooldown < System.currentTimeMillis()) {
                shoutCooldown = System.currentTimeMillis() + (getCooldownLengthInSeconds() * 1000L);
                return message;
            } else {
                long secondsLeft = (shoutCooldown - System.currentTimeMillis()) / 1000L;
                Minecraft.getMinecraft().thePlayer.addChatMessage(colorMessage("&eShout command is on cooldown. Please wait " + decimalFormat.format(secondsLeft) + " more second" + (secondsLeft == 1 ? "." : "s.")));
                return null;
            }
        }
        return message;
    }

    private long getCooldownLengthInSeconds() {
        HypixelUtils.Location location = HypixelUtils.getLocation();
        if (!"LOBBY".equals(location.getMode()) && location.getGameType().isPresent()) {
            switch (location.getGameType().get()) {
                case BEDWARS: {
                    if (!"BEDWARS_EIGHT_ONE".equals(location.getMode())) return 60L;
                    break;
                }
                case SKYWARS: {
                    return 3L;
                }
                case ARCADE: {
                    if ("PVP_CTW".equals(location.getMode().orElse(null))) return 10L;
                    break;
                }
                case UHC: {
                    if ("TEAMS".equals(location.getMode().orElse(null))) return 90L;
                    break;
                }
            }
        }
        return 0L;
    }

    @Override
    public void onMessageReceived(@NotNull ChatEvent.Receive event) {
        HypixelUtils.Location location = HypixelUtils.getLocation();
        IChatComponent message = MCTextHolder.convertToVanilla(event.getMessage());
        if (location.getGameType().orElse(null) == GameType.SKYWARS && message.getFormattedText().equals(getLanguage().cannotShoutBeforeSkywars) || // fun fact: there is no message when you shout after a skywars game
                message.getFormattedText().equals(getLanguage().cannotShoutAfterGame) ||
                message.getFormattedText().equals(getLanguage().cannotShoutBeforeGame) ||
                message.getFormattedText().equals(getLanguage().noSpectatorCommands)
        ) {
            shoutCooldown = 0L;
        }
    }

    @Override
    public boolean isEnabled() {
        return HytilsConfig.preventShoutingOnCooldown;
    }

}

File: src/main/kotlin/org/polyfrost/hytils/client/features/chat/handlers/modules/blockers/ShoutBlocker.kt
package org.polyfrost.hytils.client.features.chat.handlers.modules.blockers

import net.hypixel.data.type.GameType
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.HoverEvent
import net.minecraft.network.chat.Style
import org.polyfrost.hytils.client.HytilsRebornConfig
import org.polyfrost.hytils.client.data.providers.LanguageData
import org.polyfrost.hytils.client.events.ChatReceiveEvent
import org.polyfrost.hytils.client.events.ChatSendEvent
import org.polyfrost.hytils.client.features.chat.handlers.ChatReceiveModule
import org.polyfrost.hytils.client.features.chat.handlers.ChatSendModule
import org.polyfrost.oneconfig.api.hypixel.v1.HypixelUtils
import org.polyfrost.oneconfig.utils.v1.dsl.mc
import java.text.DecimalFormat
import kotlin.jvm.optionals.getOrNull

object ShoutBlocker : ChatSendModule, ChatReceiveModule {
    var shoutCooldown = 0L
    val decimalFormat = DecimalFormat("#.#")

    override fun onChatSend(event: ChatSendEvent) {
        if (!event.message.startsWith("/shout ")) return

        if (shoutCooldown < System.currentTimeMillis()) {
            shoutCooldown = System.currentTimeMillis() + getCooldownLengthInSeconds() * 1000
        } else {
            val secondsLeft = (shoutCooldown - System.currentTimeMillis()) / 1000L
            mc.player?.displayClientMessage(
                Component.literal("Shout command is on cooldown. Please wait ${decimalFormat.format(secondsLeft)} more second${if (secondsLeft == 1L) "" else "s"} before shouting another message.")
                    .setStyle(Style.EMPTY.withHoverEvent(
                        HoverEvent.ShowText(
                            Component.empty()
                                .append(Component.literal("Hytils Reborn\n")
                                    .withStyle(ChatFormatting.GOLD, ChatFormatting.BOLD))
                                .append(Component.literal("Your message was blocked by the \"Shout Cooldown\" setting. \nPlease wait before shouting another message.")
                                    .withStyle(ChatFormatting.GRAY))
                        )
                    )),
                false
            )
            event.cancelled = true
        }
    }

    override fun onChatReceived(event: ChatReceiveEvent) {
        val location = HypixelUtils.getLocation()

        val message = event.plainMessage
        if ((location.gameType.getOrNull() == GameType.SKYWARS && message == LanguageData.CANNOT_SHOUT_BEFORE_SKYWARS)
            || message == LanguageData.CANNOT_SHOUT_BEFORE_GAME
            || message == LanguageData.CANNOT_SHOUT_AFTER_GAME
            || message == LanguageData.NO_SPECTATOR_COMMANDS
        ) {
            shoutCooldown = 0L
        }
    }

    private fun getCooldownLengthInSeconds(): Long {
        val location = HypixelUtils.getLocation()
        if ("LOBBY" != location.mode.getOrNull() && location.gameType.isPresent) {
            when (location.gameType.get()) {
                GameType.BEDWARS -> if ("BEDWARS_EIGHT_ONE" != location.mode.getOrNull()) return 60L
                GameType.SKYWARS -> return 3L
                GameType.ARCADE -> if ("PVP_CTW" == location.mode.orElse(null)) return 10L
                GameType.UHC -> if ("TEAMS" == location.mode.orElse(null)) return 90L
                else -> {}
            }
        }
        return 0L
    }

    override val isEnabled
        get() = HytilsRebornConfig.preventShoutingOnCooldown
    override val priority = -1
}
